include/gevolution/lightcone.hpp

```diff
@@ -0,0 +1,287 @@
+        public:
+        
+        int operator()(const id_type I)
+        {
+            return I % c.size();
+        }
+    };
+    
+    distributed_database<id_type,bool,destination_func> db;
+    
+    // value of the conformal time (cosmic clock) when the lightcone was called
+    double last_tau; 
+    
+    static double sqr(double x){return x*x;}
+    
+    static bool inside_shell(const particle& part, const lightcone_geometry& shell)
+    // returns true if the particle part lies inside the lightcone shell
+    {
+        // comoving position difference
+        const std::array<double,3> pos_diff{
+            part.pos[0]-shell.vertex[0],
+            part.pos[1]-shell.vertex[1],
+            part.pos[2]-shell.vertex[2]
+        };
+        
+        // comoving distance to the observer
+        const double dist = std::sqrt(
+              sqr(pos_diff[0])
+            + sqr(pos_diff[1])
+            + sqr(pos_diff[2])
+            ) ; 
+        
+        // modulo of the direction vector
+        const double dir_mod = std::sqrt( 
+            sqr(shell.direction[0]) +
+            sqr(shell.direction[1]) +
+            sqr(shell.direction[2]) );
+        
+        const double cos_angle =  
+            ( pos_diff[0]*shell.direction[0] + 
+              pos_diff[1]*shell.direction[1] + 
+              pos_diff[2]*shell.direction[2] )
+            / dir_mod / dist;
+        
+        return dist >= shell.distance[0] && dist<shell.distance[1] && cos_angle >= shell.opening;
+    }
+    
+    public:
+    parallel_lightcone(
+          const ::boost::mpi::communicator& in_com, 
+          lightcone_geometry lc, 
+          std::string fname):
+        com{in_com}, geometry{lc}, db(com,destination_func{com}),
+        
+        // initialized to the mininum value possible of the tau (cosmic time) for an observed event
+        last_tau{std::max(0.0,geometry.tau - geometry.distance[1])}
+    {}
+   
+    // we use the conformal time (tau) as the cosmic clock, this is because tau*c is the comoving
+    // distance travelled by light and we don't need to know the cosmology to compute the shell
+    // boundaries of the lightcone. Had we use the scale factor a (or the redshift z) as our cosmic
+    // clock here, we would have to translate that into tau integrating the Hubble factor.
+    void saveLightcone(const Particles_gevolution& pcdm, 
+                       const double a /* scale factor */,
+                       const cosmology cosmo,
+                       const std::string fname)
+    {
+        const double tau = particleHorizon(a,cosmo);
+        
+        if(tau<last_tau)
+        // events with tau smaller than last_tau are not considered, they're out of the lightcone in
+        // the past
+            return;
+        
+        // at each time step we build a lightcone shell using the distance to the observer in the
+        // current and the last time step.
+        lightcone_geometry thin_shell{geometry};
```

Comment:
Comment here: the intersection of the total lightcone and the thin shell corresponding to the current time step.

Reply:
ok f5919b4ba63b8131c18b603fefd2a70d82730fea